### PR TITLE
Calendar component

### PR DESCRIPTION
**Pull request by curryy**

### CODE REVIEW

File: packages/calendar-input/src/date-input.tsx [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+      openCalendar()
+    } else if (e.keyCode === KEYS.LEFT) {
+      moveCursorLeft()
+    } else if (e.keyCode === KEYS.RIGHT) {
+      moveCursorRight()
+    } else if (e.keyCode === KEYS.HOME) {
+      setCursorPosition(0)
+      setActiveDatePart(CalendarDate.format[0])
+    } else if (e.keyCode === KEYS.END) {
+      const endPart = CalendarDate.format[CalendarDate.format.length - 1]
```
Comment (hzub) [inline packages/calendar-input/src/date-input.tsx]: It is possible to:
1) Select "day" part
2) Click "END" key
3) After input, it ends up in putting value >after< minutes:

Zrzut ekranu 2019-12-3 o 09 04 59

File: packages/calendar-input/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  getStringValueForParts = (
+    value: DateValue,
+    parts: DateTimePartsType[] = this.format
+  ) => {
+    const datePart = this.getStringValueForDateParts(value, parts)
+    const timePart = this.getStringValueForTimeParts(value, parts)
+
+    const v = `${datePart}${datePart && timePart ? ' ' : ''}${timePart}`
```
Comment (hzub) [inline packages/calendar-input/src/utils.ts]: `v` is unnecessary

File: packages/calendar-input/src/date-input.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        focusPrevPart()
+      } else if (cursorPosition > 0) {
+        setCursorPosition(cursorPosition - 1)
+      }
+    }
+  }
+
+  const handleKeyDown = (e: React.KeyboardEvent<{}>) => {
+    e.stopPropagation()
+    if (e.keyCode === KEYS.DOWN) {
```
Comment (hzub) [inline packages/calendar-input/src/date-input.tsx]: `keyCode` is being deprecated in favor of `key`: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

**Review by hzub — CHANGES_REQUESTED** [review]

File: packages/calendar-input/src/time-picker.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  minutesLabel,
+  minutesInterval,
+  hourValue,
+  minuteValue,
+  changeTime,
+  clockIcon,
+  header,
+  close,
+  isOpen,
+  collapseIcon = () => <ChevronDownIconSm />
```
[the remark this answers is not in the document]
Comment (curryy) [reply packages/calendar-input/src/time-picker.tsx]: @hzub I wonder if without chevron user will know how to exit time picker?

[the remark this answers is not in the document]
Comment (hzub) [reply packages/calendar-input/src/time-picker.tsx]: That's good thinking - but this way, we should add chevron to clock icon as well in order to stay consistent.

File: packages/calendar-input/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+export type StyledCalendarInputProps = {
+  appearance: CalendarInputAppearanceType
+}
+export type CalendarInputProps = {
+  appearance: CalendarInputAppearanceType
+  inputAppearance?: TextInputAppearanceType
+  onChange?: (date?: Date) => void
+  value?: Date
+  withTime?: boolean
```
Comment (Idered) [inline packages/calendar-input/src/types.ts]:
```suggestion
  hasTime?: boolean
```
In most components `has` prefix is used, let's keep it that way

File: stories/calendar-input.stories.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      {story()}
+    </SmashingThemeProvider>
+  ))
+  .add('appearance:default', () => <WrappedCalendar />)
+  .add('appearance:default with time', () => <WrappedCalendar withTime />)
+  .add('appearance:outline', () => (
+    <WrappedCalendar inputAppearance="outline" appearance="outline" />
+  ))
+  .add('appearance:outline with time', () => (
+    <WrappedCalendar inputAppearance="outline" withTime appearance="outline" />
```
Comment (Idered) [inline stories/calendar-input.stories.js]:
```suggestion
    <WrappedCalendar inputAppearance="outline" withTime popoverAppearance="outline" />
```

We have multiple appearance props so just `appearance` might be a little confusing. Let's change it to `popoverAppearance`

File: packages/calendar-input/src/date-input.tsx [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+      inputNode.current.childNodes.forEach(node => {
+        if (node.nodeName === 'INPUT' && activeDatePart) {
+          ;(node as HTMLInputElement).focus()
+          ;(node as HTMLInputElement).setSelectionRange(start, end)
+        }
+      })
+    }
+  }
+
+  const focusNextPart = () => {
```
Comment (Idered) [inline packages/calendar-input/src/date-input.tsx]: Would it be possible to traverse input parts using tab?

Comment (curryy) [reply packages/calendar-input/src/date-input.tsx]: Yes, you can do it now

**Review by Idered — APPROVED** [review]
Splendid work 👏 

I guess that it's good enough to merge. There are some minor issues that can be resolved in other PR.

**Review by hzub — APPROVED** [review]

**Pull request merged**